extensions/godslayerakp/ws.js

```diff
@@ -0,0 +1,283 @@
+          },
+          {
+            opcode: "isClosed",
+            blockType: BlockType.BOOLEAN,
+            text: "is connection closed?",
+          },
+          {
+            opcode: "closeCode",
+            blockType: BlockType.REPORTER,
+            disableMonitor: true,
+            text: "closing code",
+          },
+          {
+            opcode: "closeMessage",
+            blockType: BlockType.REPORTER,
+            disableMonitor: true,
+            text: "closing message",
+          },
+          {
+            opcode: "closeWithoutReason",
+            blockType: BlockType.COMMAND,
+            text: "close connection without reason",
+          },
+          {
+            opcode: "closeWithCode",
+            blockType: BlockType.COMMAND,
+            arguments: {
+              CODE: {
+                type: ArgumentType.NUMBER,
+                defaultValue: "1000",
+              },
+            },
+            text: "close connection with code [CODE]",
+          },
+          {
+            opcode: "closeWithReason",
+            blockType: BlockType.COMMAND,
+            arguments: {
+              CODE: {
+                type: ArgumentType.NUMBER,
+                defaultValue: "1000",
+              },
+              REASON: {
+                type: ArgumentType.STRING,
+                defaultValue: "connection fulfilled",
+              },
+            },
+            text: "close connection with reason [REASON] and code [CODE]",
+          },
+        ],
+      };
+    }
+    async newInstance(args, utils) {
+      let url = Cast.toString(args.URL);
+      if (!url.startsWith("ws")) {
+        // url doesnt start with a valid connection type
+        // so we jsut assume its formated without it
+        if (/^(?!(ws|http)s?:\/\/).*$/is.test(url)) {
+          url = `wss://${url}`;
+        } else {
+          if (!url.startsWith("http")) return;
+          url = url
+            .split(":")
+            // @ts-ignore
+            .with(0, url.startsWith("https") ? "wss" : "ws")
+            .join(":");
+        }
+      }
+      if (!(await Scratch.canFetch(url))) return;
+
+      try {
+        // eslint-disable-next-line no-restricted-syntax
+        const websocket = new WebSocket(url);
+        const instance = {
+          closeMessage: "",
+          closeCode: "",
+          data: "",
+          isOpen: false,
+          isClosed: false,
+          isErrored: false,
+          isMessage: false,
+          set gottenMessage(bool) {
+            this.isMessage = bool;
+          },
+          get gottenMessage() {
+            const oldState = this.isMessage;
+            this.isMessage = false;
+            return oldState;
+          },
+          state: "opening",
+          websocket,
+        };
+        websocket.onopen = (e) => {
+          instance.state = "opened";
+          instance.isOpen = true;
+          utils.startHats("websocket_onOpen", null, utils.target);
```

Comment:
The only safe time to use the block utility is immediately, in a syncronous script. You should use runtime.startHats instead

Reply:
Similarly, you need to store `utils.target` in a local variable somewhere because the object gets reused for every script

Reply:
> Similarly, you need to store `utils.target` in a local variable somewhere because the object gets reused for every script, so right now none of these events work if any other sprites have running scripts

oh...

Reply:
(utils.startHats is actually probably safe to use, but strictly you shouldnt really be touching it at all at this point, so try not to use it)